broker.go

```diff
@@ -0,0 +1,584 @@
+	}
+	return syncDone, true
+}
+
+// onEvent is called by the datatransfer manager to send events.
+func (lb *LegBroker) onEvent(event dt.Event, channelState dt.ChannelState) {
+	if event.Code != dt.FinishTransfer {
+		return
+	}
+	// Find the channel to notify the handler that the transfer is finished.
+	syncDone, ok := lb.popSyncDoneChan(channelState.BaseCID())
+	if !ok {
+		log.Errorw("could not find channel for completed transfer notice", "cid", channelState.BaseCID())
+		return
+	}
+
+	// Send the FinishTransfer signal to the handler.  This will allow its
+	// handle goroutine to distribute the update and exit.
+	close(syncDone)
+}
+
+// handleAsync starts a goroutine to process the latest message received over pubsub.
+func (h *handler) handleAsync(ctx context.Context, c cid.Cid, ss ipld.Node) {
+	// Remove any previous message and replace it with the most recent.
+	select {
+	case oldmsg := <-h.msgChan:
+		log.Info("Message %s replaced by %s", oldmsg, c)
+	default:
+	}
+
+	// Put new message on channel.  This is necessary so that if multiple
+	// messages arrive while this handler is already syncing, that messages are
+	// handled in order, regardless of goroutine scheduling.
+	h.msgChan <- c
+
+	go func() {
+		// Wait for any previous messages to be handled.
+		h.syncMutex.Lock()
+		defer h.syncMutex.Unlock()
+
+		select {
+		case <-ctx.Done():
+		case c := <-h.msgChan:
+			err := h.handle(ctx, c, ss, true, true)
+			if err != nil {
+				// Log error for now.
+				log.Errorw("Cannot process message", "err", err, "peer", h.peerID)
+			}
+		default:
+			// A previous goroutine, that had its message replaced, read the
+			// message.  Or, the message was removed and will be replaced by
+			// another message and goroutine.  Either way, nothing to do in this
+			// routine.
+		}
+	}()
+}
+
+// handle processes a message from the peer that the handler is responsible.
+// The caller is responsible for ensuring that this is called while h.syncMutex
+// is locked.
+func (h *handler) handle(ctx context.Context, c cid.Cid, selSeq ipld.Node, wrapSel, updateLatest bool) error {
+	v := Voucher{&c}
+
+	log.Debugw("Starting data channel for message source", "cid", c, "latest_sync", h.latestSync, "source_peer", h.peerID)
+
+	syncDone := make(chan struct{})
+	h.putSyncDone(c, syncDone)
+
+	if wrapSel {
+		selSeq = ExploreRecursiveWithStopNode(selector.RecursionLimitNone(), selSeq, h.latestSync)
+	}
+	_, err := h.dtManager.OpenPullDataChannel(ctx, h.peerID, &v, c, selSeq)
+	if err != nil {
+		return fmt.Errorf("cannot open data channel: %s", err)
+		// TODO: Should we retry if there's an error in the exchange?
+	}
+
+	// Wait for transfer finished signal.
+	<-syncDone
+
+	if !updateLatest {
+		return nil
+	}
+
+	// Update latest head seen.
+	// NOTE: This is not persisted anywhere. Is the top-level user's
+	// responsibility to persist if needed to initialize a
+	// partiallySynced subscriber.
+	h.latestSync = cidlink.Link{Cid: c}
```

Comment:
Can I ask you to document why we are not locking the `syncMutex` , and why are we not using `setLatestSync` instead please.